src/circuit/dialer.js

```
@@ -0,0 +1,238 @@
+'use strict'
+
+const Connection = require('interface-connection').Connection
+const isFunction = require('lodash.isfunction')
+const multiaddr = require('multiaddr')
+const once = require('once')
+const waterfall = require('async/waterfall')
+const utilsFactory = require('./utils')
+const StreamHandler = require('./stream-handler')
+const PeerId = require('peer-id')
+
+const debug = require('debug')
+const log = debug('libp2p:circuit:dialer')
+log.err = debug('libp2p:circuit:error:dialer')
+
+const multicodec = require('../multicodec')
+const proto = require('../protocol')
+
+class Dialer {
+  /**
+   * Creates an instance of Dialer.
+   * @param {Swarm} swarm - the swarm
+   * @param {any} options - config options
+   *
+   * @memberOf Dialer
+   */
+  constructor (swarm, options) {
+    this.swarm = swarm
+    this.relayPeers = new Map()
+    this.options = options
+    this.utils = utilsFactory(swarm)
+  }
+
+  /**
+   * Dial a peer over a relay
+   *
+   * @param {multiaddr} ma - the multiaddr of the peer to dial
+   * @param {Function} cb - a callback called once dialed
+   * @returns {Connection} - the connection
+   *
+   * @memberOf Dialer
+   */
+  dial (ma, cb) {
+    const strMa = ma.toString()
+    if (!strMa.includes('/p2p-circuit')) {
+      log.err('invalid circuit address')
+      throw new Error('invalid circuit address')
```

Should return the error in the callback here